_CS141/part2.md

```diff
@@ -0,0 +1,308 @@
+
+**Example.**
+
+```haskell
+fac' :: Int -> Int -> Int
+fac' 0 m = m
+fac' n m = fac' (n-1) (n*m)
+```
+
+```haskell
+   fac 2
+=> fac' 2 1
+=> fac' (2-1) (2*1)
+=> fac' (1) (2*1)
+=> fac' 1 2
+=> fac' (1-1) (1*2)
+   ...
+```
+
+### Call-by-name
+
+> A **non-strict evaluation strategy** where expressions are given to functions as arguments are not reduced before the function call is made. 
+>
+> Expressions are only reduced when their value is **needed**.
+
+**When is a value needed?**
+
+A case expression is the only thing that enforces the evaluation of a particular expression (we’re only evaluating something if its in a case expression). We **need the value** of an expression when we **cannot proceed** with the case expression **until we reduce** the expression.
+
+```haskell
+fac' :: Int -> Int -> Int
+fac' 0 m = m
+fac' n m = fac' (n-1) (n*m)
+```
+
+```haskell
+   fac 2
+=> fac' 2 1
+=> case 2 of
+     0 -> 1
+     _ -> fac' (2-1) (2*1)
+=> fac' (2-1) (2*1)
+=> case 2-1 of
+     0 -> 2*1
+     _ -> fac' ((2-1)-1) ((2-1)*(2*1))
+=> case 1 of
+     0 -> 2*1
+     _ -> fac' ((2-1)-1) ((2-1)*(2*1))
+```
+
+In **call-by-name**, the 2nd last `=>` to the last `=>` is the only time an expression was evaluated, as the value of the redex `2-1` was needed to continue.
+
+### CBN vs CBV
+
+| Call-by-value                                                | Call-by-name                                                 |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| Reduce function arguments to normal forms before calling the function | Only reduce expressions when their value is needed           |
+| We may evaluate arguments even if they are **never needed**. | We may end up reducing the same expression **over and over**. |
+
+Both have flaws. Haskell’s fix is to use **lazy evaluation.**
+
+### Lazy evaluation
+
+> This is essentially **call-by-name + sharing**, where we avoid duplicate evaluation but also only ever evaluate expressions that are needed.
+
+Lazy evaluation is the default evaluation strategy in Haskell, sometimes it can be strict, if the **compiler** (the compiler does everything for us 🥳) thinks its better. 
+
+#### Sharing
+
+This is the technique that helps us avoid duplicate evaluation.
+
+> Sharing turns arguments to function into local definitions. 
+
+```haskell
+                              fac' n m = case n of
+fac' n m = case n of            0 -> m
+  0 -> m                 ==>    _ -> let x = n - 1
+  _ -> fac' (n-1) (n*m)                  y = n * m
+                                     in fac' x y
+```
+
+This ensures that functions are always applied to either **values** or a **variable** defined in a `let` (or `where`) bound. This means that if a variable (e.g. `x0 = 2-1`) has to be evaluated, its RHS is evaluated (so `2-1=1`) and `x0` is updated with the value of the expression, so `x0 = 1`. 
```

Comment:
A suitable high-level explanation. Might be worth highlighting the correspondence between let/where bindings to dynamic closures somewhere.